python/ray/dag/compiled_dag_node.py

```diff
@@ -946,13 +968,55 @@ def _preprocess(self) -> None:
                     # Add all readers to the NCCL group.
                     nccl_actors.add(downstream_actor_handle)
 
-        # If there were type hints indicating transport via NCCL, initialize
-        # the NCCL group on the participating actors.
         nccl_actors = list(nccl_actors)
         if None in nccl_actors:
             raise ValueError("Driver cannot participate in the NCCL group.")
-        if nccl_actors and self._nccl_group_id is None:
+
+        # Initialize a NCCL group for each set of actors. A set of actors can be
+        # calling P2P send/recv or collective methods.
+        actors_to_nccl_group_id: Dict[FrozenSet["ray.actor.ActorHandle"], str] = {}
+
+        # If a custom NCCL group is specified for P2P actors, initialize and cache
+        # the NCCL group ID.
+        if nccl_actors and self._custom_nccl_group:
             self._nccl_group_id = _init_nccl_group(nccl_actors, self._custom_nccl_group)
```

Comment:
Can you add a comment and explain `self._nccl_group_id`, now that there are multiple nccl_group_ids

Reply:
Done.